File: cawbs/Dart/lib/wbs.dart
// Copyright Core DF

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Shared HTTP helpers for the Core Auto Collector (cawbs) Dart client.

import 'dart:convert';
import 'package:http/http.dart' as http;

class WbsResult {
  final int statusCode;
  final dynamic error;
  final dynamic payload;
  final Map<String, dynamic>? answer;

  WbsResult({
    required this.statusCode,
    this.error,
    this.payload,
    this.answer,
  });

  Map<String, dynamic> toMap() => {
        'status_code': statusCode,
        if (error != null) 'error': error,
        if (payload != null) 'payload': payload,
        if (answer != null) 'answer': answer,
      };
}

class WbsSession {
  bool _initialized = false;
  String _baseUrl = '';
  String _env = '';
  String _token = '';

  static WbsResult missingEnv(String vars) => WbsResult(
        statusCode: 601,
        error: 'Environment variables $vars should be defined',
      );

  String _trimUrl(String url) => url.replaceAll(RegExp(r'^[/ ]+|[/ ]+$'), '');

  Future<({int statusCode, dynamic body, bool transportError})> _request(
    String method,
    String url, {
    String? body,
  }) async {
    final headers = <String, String>{
      'Content-Type': 'application/json',
      'Environment': _env,
    };
    if (_token.isNotEmpty) {
      headers['Authorization'] = 'Bearer $_token';
    }
    try {
      late http.Response resp;
      if (method == 'GET') {
        resp = await http.get(Uri.parse(url), headers: headers);
      } else {
        resp = await http.post(Uri.parse(url), headers: headers, body: body);
      }
      dynamic parsed;
      if (resp.body.isNotEmpty) {
        try {
          parsed = jsonDecode(resp.body);
        } catch (_) {
          parsed = null;
        }
      }
      return (statusCode: resp.statusCode, body: parsed, transportError: false);
    } catch (_) {
      return (statusCode: 0, body: null, transportError: true);
    }
  }

  WbsResult _apiError(int statusCode, dynamic body) => WbsResult(
        statusCode: statusCode,
        error: body ?? 'inaccessible',
      );

  Future<WbsResult> authenticate(String env, String accessCode, String baseUrl) async {
    if (_initialized) {
      return WbsResult(statusCode: 602, error: 'init already called');
    }
    _env = env;
    _baseUrl = _trimUrl(baseUrl);
    final out = await _request(
      'POST',
      '$_baseUrl/v1/auth/apicode',
      body: jsonEncode({'apiCode': accessCode}),
    );
    if (out.transportError) {
      return WbsResult(statusCode: out.statusCode, error: 'inaccessible');
    }
    if (out.statusCode >= 400) return _apiError(out.statusCode, out.body);
    final token = (out.body as Map?)?['token'];
    if (token is! String || token.isEmpty) {
      return WbsResult(statusCode: out.statusCode, error: 'inaccessible');
    }
    _token = token;
    _initialized = true;
    return WbsResult(statusCode: out.statusCode);
  }

  Future<WbsResult> getEventPayload(String actionId) async {
    if (!_initialized) return WbsResult(statusCode: 603, error: 'Init required');
    final out = await _request('GET', '$_baseUrl/v1/rtevent/$actionId');
    if (out.transportError) return WbsResult(statusCode: out.statusCode, error: 'inaccessible');
    if (out.statusCode >= 400) return _apiError(out.statusCode, out.body);
    if (out.body is! Map) return WbsResult(statusCode: out.statusCode, error: 'inaccessible');
    return WbsResult(statusCode: out.statusCode, payload: out.body['payload']);
  }

  Future<WbsResult> putStepPayload(String actionId, String stepName, dynamic payload) async {
    if (!_initialized) return WbsResult(statusCode: 603, error: 'Init required');
    final out = await _request(
      'POST',
      '$_baseUrl/v1/rtstep/payload',
      body: jsonEncode({'actionId': actionId, 'stepname': stepName, 'payload': payload}),
    );
    if (out.transportError) return WbsResult(statusCode: out.statusCode, error: 'inaccessible');
    if (out.statusCode >= 400) return _apiError(out.statusCode, out.body);
    return WbsResult(statusCode: out.statusCode);
  }

  Future<WbsResult> getStepPayload(String actionId, String stepName) async {
    if (!_initialized) return WbsResult(statusCode: 603, error: 'Init required');
    final out = await _request('GET', '$_baseUrl/v1/rtstep/payload/$actionId/$stepName');
    if (out.transportError) return WbsResult(statusCode: out.statusCode, error: 'inaccessible');
    if (out.statusCode >= 400) return _apiError(out.statusCode, out.body);
    if (out.body is! Map) return WbsResult(statusCode: out.statusCode, error: 'inaccessible');
    return WbsResult(statusCode: out.statusCode, payload: out.body['payload']);
  }

  Future<WbsResult> getKeystore(String keylist) async {
    if (!_initialized) return WbsResult(statusCode: 603, error: 'Init required');
    final keys = keylist.replaceAll(' ', '');
    final out = await _request('GET', '$_baseUrl/v1/keystore/$keys');
    if (out.transportError) return WbsResult(statusCode: out.statusCode, error: 'inaccessible');
    if (out.statusCode >= 400) return _apiError(out.statusCode, out.body);
    if (out.body is! Map<String, dynamic>) {
      return WbsResult(statusCode: out.statusCode, error: 'inaccessible');
    }
    for (final key in keys.split(',')) {
      if (key.isEmpty) continue;
      if (!out.body.containsKey(key)) {
        return WbsResult(statusCode: 605, error: '$key not found');
      }
    }
    return WbsResult(statusCode: out.statusCode, answer: Map<String, dynamic>.from(out.body));
  }
}

File: cawbs/Ruby/lib/wbs.rb
# frozen_string_literal: true

# Copyright Core DF

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# Shared HTTP helpers for the Core Auto Collector (cawbs) Ruby client.

require 'json'
require 'net/http'
require 'uri'

module Wbs
  Result = Struct.new(:status_code, :error, :payload, :answer, keyword_init: true) do
    def to_h
      h = { status_code: status_code }
      h[:error] = error unless error.nil?
      h[:payload] = payload unless payload.nil?
      h[:answer] = answer unless answer.nil?
      h
    end
  end

  module_function

  def missing_env(vars)
    Result.new(status_code: 601, error: "Environment variables #{vars} should be defined")
  end

  def trim_url(url)
    url.gsub(/\A[\/ ]+|[\/ ]+\z/, '')
  end

  def do_request(method, url, headers, body = nil)
    uri = URI(url)
    klass = method == :get ? Net::HTTP::Get : Net::HTTP::Post
    req = klass.new(uri)
    headers.each { |k, v| req[k] = v }
    req.body = body if body

    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = uri.scheme == 'https'
    http.open_timeout = 60
    http.read_timeout = 60

    resp = http.request(req)
    parsed = nil
    if resp.body && !resp.body.empty?
      begin
        parsed = JSON.parse(resp.body)
      rescue JSON::ParserError
        parsed = nil
      end
    end
    [resp.code.to_i, parsed]
  rescue StandardError
    [0, nil]
  end

  def api_error(status_code, body)
    return Result.new(status_code: status_code, error: 'inaccessible') if body.nil?

    Result.new(status_code: status_code, error: body)
  end

  class Session
    attr_reader :initialized

    def initialize
      @initialized = false
      @base_url = ''
      @headers = {}
    end

    def authenticate(env, access_code, base_url)
      return Result.new(status_code: 602, error: 'init already called') if @initialized

      @base_url = Wbs.trim_url(base_url)
      headers = {
        'Content-Type' => 'application/json',
        'Environment' => env
      }
      status_code, body = Wbs.do_request(
        :post,
        "#{@base_url}/v1/auth/apicode",
        headers,
        JSON.generate(apiCode: access_code)
      )
      return Result.new(status_code: status_code, error: 'inaccessible') if status_code.zero?
      return Wbs.api_error(status_code, body) if status_code >= 400
      return Result.new(status_code: status_code, error: 'inaccessible') unless body.is_a?(Hash) && body['token']

      @headers = headers.merge('Authorization' => "Bearer #{body['token']}")
      @initialized = true
      Result.new(status_code: status_code)
    end

    def get_event_payload(action_id)
      return Result.new(status_code: 603, error: 'Init required') unless @initialized

      status_code, body = Wbs.do_request(:get, "#{@base_url}/v1/rtevent/#{action_id}", @headers)
      return Result.new(status_code: status_code, error: 'inaccessible') if status_code.zero?
      return Wbs.api_error(status_code, body) if status_code >= 400
      return Result.new(status_code: status_code, error: 'inaccessible') if body.nil?

      Result.new(status_code: status_code, payload: body['payload'])
    end

    def put_step_payload(action_id, step_name, payload)
      return Result.new(status_code: 603, error: 'Init required') unless @initialized

      body_json = JSON.generate(actionId: action_id, stepname: step_name, payload: payload)
      status_code, body = Wbs.do_request(:post, "#{@base_url}/v1/rtstep/payload", @headers, body_json)
      return Result.new(status_code: status_code, error: 'inaccessible') if status_code.zero?
      return Wbs.api_error(status_code, body) if status_code >= 400

      Result.new(status_code: status_code)
    end

    def get_step_payload(action_id, step_name)
      return Result.new(status_code: 603, error: 'Init required') unless @initialized

      status_code, body = Wbs.do_request(
        :get,
        "#{@base_url}/v1/rtstep/payload/#{action_id}/#{step_name}",
        @headers
      )
      return Result.new(status_code: status_code, error: 'inaccessible') if status_code.zero?
      return Wbs.api_error(status_code, body) if status_code >= 400
      return Result.new(status_code: status_code, error: 'inaccessible') if body.nil?

      Result.new(status_code: status_code, payload: body['payload'])
    end

    def get_keystore(keylist)
      return Result.new(status_code: 603, error: 'Init required') unless @initialized

      keys = keylist.gsub(' ', '')
      status_code, body = Wbs.do_request(:get, "#{@base_url}/v1/keystore/#{keys}", @headers)
      return Result.new(status_code: status_code, error: 'inaccessible') if status_code.zero?
      return Wbs.api_error(status_code, body) if status_code >= 400
      return Result.new(status_code: status_code, error: 'inaccessible') if body.nil?

      keys.split(',').each do |key|
        next if key.empty?
        return Result.new(status_code: 605, error: "#{key} not found") unless body.key?(key)
      end
      Result.new(status_code: status_code, answer: body)
    end
  end
end
